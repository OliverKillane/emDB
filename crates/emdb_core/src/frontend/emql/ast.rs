//! # emQL Abstract Syntax Tree
//! A basic AST, without error nodes. Includes spans required for [`super::sem`] to create [`super::errors`].

use super::operators::Operator;
use proc_macro2::{Ident, Span, TokenStream};
use syn::{Expr, Type};

#[derive(Debug)]
pub(super) enum AstType {
    RsType(syn::Type),
    TableRef(Ident),
    Custom(Ident),
}

#[derive(Debug)]
pub(super) struct Connector {
    /// single (~>) or stream (|>)
    pub stream: bool,
    pub span: Span,
}

#[derive(Debug)]
pub(super) struct StreamExpr {
    pub op: Operator,
    pub con: Option<(Connector, Box<StreamExpr>)>,
}

#[derive(Debug)]
pub(super) enum ConstraintExpr {
    Unique { field: Ident },
    Pred(Expr),
    Limit { size: Expr },
}

#[derive(Debug)]
pub(super) struct Constraint {
    pub alias: Ident,
    pub method_span: Span,
    pub expr: ConstraintExpr,
}

#[derive(Debug)]
pub(super) struct Table {
    pub name: Ident,
    pub cols: Vec<(Ident, Type)>,
    pub cons: Vec<Constraint>,
}

#[derive(Debug)]
pub struct Context {
    pub params: Vec<(Ident, AstType)>,
    pub streams: Vec<StreamExpr>,
}

#[derive(Debug)]
pub(super) struct Query {
    pub name: Ident,
    pub context: Context,
}

#[derive(Debug)]
pub(super) struct BackendImpl {
    pub impl_name: Ident,
    pub backend_name: Ident,
    pub options: Option<TokenStream>,
}

#[derive(Debug)]
pub(super) struct Ast {
    pub backends: Vec<BackendImpl>,
    pub tables: Vec<Table>,
    pub queries: Vec<Query>,
}
