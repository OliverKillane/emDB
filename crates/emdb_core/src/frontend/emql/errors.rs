//! # EmQL Error Messages
//! The general error messages produced by emql's semantic analysis (from [`super::sem`] and [`super::operators`])
//! - Each error has a code identified (for easy communication/bug reports)

use super::sem::VarState;
use crate::plan::{self, Key, Plan, RecordType, ScalarType, Table, With};
use itertools::Itertools;
use proc_macro2::{Ident, Span};
use proc_macro_error::{Diagnostic, Level};
use std::collections::HashMap;
use syn::Type;

type ErrCode = usize;

fn emql_error(code: ErrCode, span: Span, message: String) -> Diagnostic {
    Diagnostic::spanned(span, Level::Error, format!("[EMQL-{code}] {message}"))
}

fn redefinition_error(
    err_code: ErrCode,
    def_type: &str,
    def: &Ident,
    original_def: &Ident,
) -> Diagnostic {
    emql_error(
        err_code,
        def.span(),
        format!("Redefinition of {def_type} `{def}`"),
    )
    .span_note(original_def.span(), "Originally defined here".to_string())
    .help(format!("Each {def_type} must have a unique name"))
}

pub fn backend_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(0, "backend", def, original_def)
}

pub fn table_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(1, "table", def, original_def)
}

pub fn table_column_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(2, "table column", def, original_def)
}

pub fn table_constraint_alias_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(3, "constraint alias", def, original_def)
}

pub fn collect_type_alias_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(4, "collect type alias", def, original_def)
}

pub fn table_constraint_duplicate_unique(
    col_name: &Ident,
    method_span: Span,
    prev_alias: &Ident,
) -> Diagnostic {
    emql_error(
        5,
        method_span,
        format!("Duplicate unique constraint on column `{col_name}`"),
    )
    .span_note(
        prev_alias.span(),
        format!("previously defined as {prev_alias} here."),
    )
}

pub fn table_constraint_nonexistent_unique_column(
    alias: &Ident,
    col_name: &Ident,
    table_name: &Ident,
    method_span: Span,
) -> Diagnostic {
    emql_error(6, method_span, format!(
        "Column `{col_name}` does not exist in table `{table_name}`, so cannot apply a unique constraint `{alias}` to it"
    )).span_help(table_name.span(), format!("Apply the unique constraint to an available column in {table_name}"))
}

pub fn table_constraint_duplicate_limit(
    alias: &Ident,
    table_name: &Ident,
    method_span: Span,
) -> Diagnostic {
    emql_error(
        7,
        method_span,
        format!("Duplicate limit constraint `{alias}` on table `{table_name}`"),
    )
    .span_help(
        table_name.span(),
        "Limit constraints can only be applied once".to_string(),
    )
}

pub fn query_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(8, "query", def, original_def)
}

pub fn query_multiple_returns(ret: Span, prev_ret: Span, query: &Ident) -> Diagnostic {
    emql_error(
        9,
        ret,
        format!("Multiple return statements in query `{query}`"),
    )
    .span_help(prev_ret, "Previously returned value here".to_string())
}

pub fn query_operator_field_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(10, "field", def, original_def)
}

pub fn query_stream_single_connection(span: Span, last_span: Span, stream: bool) -> Diagnostic {
    emql_error(
        11,
        span,
        String::from(if stream {
            "Expected a stream, but found a single connector"
        } else {
            "Expected a single, but found a stream connector"
        }),
    )
    .span_note(
        last_span,
        "The previous operator provides the values".to_string(),
    )
    .help(
        if stream {
            "Use a stream connector `|>` instead of single `~>`"
        } else {
            "Use a single connector `~>` instead of stream `|>`"
        }
        .to_string(),
    )
}

pub fn query_no_data_for_next_operator(conn_span: Span, stream: bool, prev_op: Span) -> Diagnostic {
    emql_error(
        12,
        prev_op,
        "No output data provided for next operator".to_string(),
    )
    .span_note(
        conn_span,
        format!(
            "Expected a {} out here",
            if stream {
                "stream of data"
            } else {
                "singe data"
            }
        ),
    )
}

pub fn query_early_return(conn_span: Span, stream: bool, ret_op: Span) -> Diagnostic {
    emql_error(13, ret_op, "Early return statement".to_string()).span_note(
        conn_span,
        format!(
            "Expected a {} out here",
            if stream {
                "stream of data (`|>`)"
            } else {
                "single data (`~>`)"
            }
        ),
    )
}

pub fn query_parameter_redefined(def: &Ident, original_def: &Ident) -> Diagnostic {
    redefinition_error(14, "query parameter", def, original_def)
}

pub fn query_param_ref_table_not_found(query: &Ident, table_ref: &Ident) -> Diagnostic {
    emql_error(
        15,
        table_ref.span(),
        format!("Table `{table_ref}` not found in query `{query}`"),
    )
    .help(format!(
        "Either use a `ref <other table>` or create `table {table_ref} {{...}} @ [...]"
    ))
}

pub fn access_field_missing(call: &Ident, field: &Ident, fields: Vec<&Ident>) -> Diagnostic {
    emql_error(
        16,
        field.span(),
        format!(
            "`{call}` field `{field}` is not found, the available fields are {}",
            fields.iter().join(", ")
        ),
    )
}

pub fn query_expected_reference_type_for_update(
    lp: &Plan,
    dt: &Key<ScalarType>,
    reference: &Ident,
) -> Diagnostic {
    emql_error(
        17,
        reference.span(),
        format!(
        "Expected a reference to a table for the update in `{reference}`, but got a `{}` instead",
        With { plan: lp, extended: dt }
    ),
    )
    .help(format!(
        "Assign a reference to {reference}, or use a different field that is a reference"
    ))
}

pub fn query_cannot_start_with_operator(op: &Ident) -> Diagnostic {
    emql_error(
        18,
        op.span(),
        format!("Cannot start a query expression with operator `{op}`"),
    )
    .help("Instead use operators such as `use ..`, `ref ..`, or `unique(..)`".to_string())
}

pub fn query_update_field_not_in_table(table_name: &Ident, field: &Ident) -> Diagnostic {
    emql_error(
        19,
        field.span(),
        format!("Field `{field}` not found in table `{table_name}` and hence cannot be updated",),
    )
    .span_note(
        table_name.span(),
        format!("The fields that can be updated are present in the `{table_name}` definition"),
    )
}

pub fn query_update_reference_not_present(
    lp: &Plan,
    reference: &Ident,
    prev_span: Span,
    dt: &Key<RecordType>,
) -> Diagnostic {
    emql_error(
        20,
        reference.span(),
        format!("Reference `{reference}` not found in the query"),
    )
    .span_note(
        prev_span,
        format!(
            "The previous operator produces {}",
            With {
                plan: lp,
                extended: dt
            }
        ),
    )
}

pub fn query_insert_field_rust_type_mismatch(
    lp: &Plan,
    call: &Ident,
    field: &Ident,
    passed_type: &Type,
    expected_type: &Type,
    prev_span: Span,
) -> Diagnostic {
    emql_error(21, call.span(), format!(
        "Field `{field}` has type `{passed_type:#?}` which does not match the expected type `{expected_type:#?}`"
    )).span_note(field.span(), format!("`{field}` defined here"))
    .span_note(prev_span, format!("Input to `{call}` comes from here"))
}

pub fn query_insert_field_type_mismatch(
    lp: &Plan,
    call: &Ident,
    field: &Ident,
    passed_type: &Key<RecordType>,
    expected_type: &Type,
    prev_span: Span,
) -> Diagnostic {
    emql_error(
        22,
        call.span(),
        format!(
            "Field `{field}` has type `{}` which does not match the expected type `{:#?}`",
            With {
                plan: lp,
                extended: passed_type
            },
            expected_type
        ),
    )
    .span_note(field.span(), format!("`{field}` defined here"))
    .span_note(prev_span, format!("Input to `{call}` comes from here"))
}

pub fn query_insert_field_missing(
    call: &Ident,
    table_name: &Ident,
    field: &Ident,
    last_span: Span,
) -> Diagnostic {
    emql_error(
        23,
        call.span(),
        format!("Field `{field}` is missing from the insert into table `{table_name}`"),
    )
    .span_note(field.span(), format!("`{field}` defined here"))
    .span_error(
        last_span,
        format!("`{field}` should have been part of the output from here"),
    )
}

pub fn query_insert_extra_field(call: &Ident, field: &Ident, table_name: &Ident) -> Diagnostic {
    emql_error(
        24,
        call.span(),
        format!("Field `{field}` is not a valid field for table `{table_name}`"),
    )
    .span_note(field.span(), format!("`{field}` defined here"))
    .span_note(table_name.span(), format!("`{table_name}` defined here"))
}

pub fn query_nonexistent_table(call: &Ident, table_used: &Ident) -> Diagnostic {
    emql_error(25, table_used.span(), format!(
        "Table `{table_used}` does not exist in the query so cannot be accessed through `{call}`",
    )).help(format!(
        "Either define a `table {table_used} {{...}} @ [...]` or use a different table in `insert(..)`", 
    ))
}

pub fn query_delete_field_not_present(call: &Ident, field: &Ident) -> Diagnostic {
    emql_error(
        26,
        field.span(),
        format!("Field `{field}` not found in the available data"),
    )
    .span_help(
        call.span(),
        format!("`{field}` needs to be accessible here"),
    )
}

pub fn query_delete_field_not_reference(
    lp: &Plan,
    call: &Ident,
    field: &Ident,
    dt: &Key<ScalarType>,
) -> Diagnostic {
    emql_error(
        27,
        field.span(),
        format!(
            "Field `{field}` is not a reference, but a `{}`",
            With {
                plan: lp,
                extended: dt
            }
        ),
    )
    .span_help(call.span(), format!("`{field}` "))
}

pub fn query_deref_field_already_exists(new: &Ident, existing: &Ident) -> Diagnostic {
    emql_error(28, new.span(), format!("Field `{new}` already exists"))
        .span_note(existing.span(), format!("{existing} defined here"))
        .help(format!(
            "Rename `{new}` or remove the existing `{new}` field"
        ))
}

pub fn query_reference_field_missing(reference: &Ident) -> Diagnostic {
    emql_error(
        29,
        reference.span(),
        format!("Field `{reference}` not found in the available data"),
    )
    .span_help(
        reference.span(),
        format!("`{reference}` needs to be accessible here"),
    )
}

pub fn query_deref_cannot_deref_rust_type(reference: &Ident, t: &Type) -> Diagnostic {
    emql_error(
        30,
        reference.span(),
        format!("Cannot dereference a rust type `{t:#?}`"),
    )
}

pub fn query_deref_cannot_deref_record(
    lp: &Plan,
    reference: &Ident,
    t: &Key<RecordType>,
) -> Diagnostic {
    emql_error(
        31,
        reference.span(),
        format!(
            "Cannot dereference a record `{}`",
            With {
                plan: lp,
                extended: t
            }
        ),
    )
}

pub fn query_operator_cannot_come_first(call: &Ident) -> Diagnostic {
    emql_error(
        32,
        call.span(),
        format!("Operator `{call}` cannot be the first operator in a query"),
    )
}

pub fn query_unique_table_not_found(table: &Ident) -> Diagnostic {
    emql_error(
        33,
        table.span(),
        format!("Table `{table}` not found in the query"),
    )
    .help(format!(
        "Either define a `table {table} {{...}} @ [...]` or use a different table in `unique(..)`",
    ))
}

pub fn query_unique_no_field_in_table(field: &Ident, table_name: &Ident) -> Diagnostic {
    emql_error(
        34,
        field.span(),
        format!("Field `{field}` not found in table `{table_name}`"),
    )
    .span_help(
        table_name.span(),
        format!("Add `{field}: ... ,` to {table_name}"),
    )
}

pub fn query_unique_field_is_not_unique(field: &Ident, table_name: &Ident) -> Diagnostic {
    emql_error(35, field.span(), format!("Field `{field}` is not unique in table `{table_name}`"))
    .span_help(
        table_name.span(),
        format!(
            "Add a unique constraint to `{field}` in {table_name} `@ [ ... unique({field}) as ... ]`"
        ),
    )
}

pub fn query_use_variable_already_used(usage: &Ident, created: Span, used: Span) -> Diagnostic {
    emql_error(
        36,
        usage.span(),
        format!("Variable `{usage}` has already been used"),
    )
    .span_error(created, "Was created here".to_string())
    .span_error(used, "And consumed here".to_string())
}

pub fn query_invalid_use(
    usage: &Ident,
    tn: &HashMap<Ident, Key<Table>>,
    vs: &HashMap<Ident, VarState>,
) -> Diagnostic {
    let vars = vs
        .iter()
        .filter_map(|(var, state)| {
            if matches!(state, VarState::Available { .. }) {
                Some(var)
            } else {
                None
            }
        })
        .join(", ");
    let tables = tn.keys().join(", ");
    emql_error(37,usage.span(), format!("Invalid use of variable `{usage}`",))
    .help(format!("Currently available variables are {vars}, and tables {tables}"  ))
    .help(format!(
        "To introduce a new `{usage}` make a new table `table {usage} {{ ... }} @ [ ... ]` or a new variable ` ... |> let {usage}`"
    ))
}

pub fn query_invalid_variable_use(usage: &Ident, vs: &HashMap<Ident, VarState>) -> Diagnostic {
    let vars = vs
        .iter()
        .filter_map(|(var, state)| {
            if matches!(state, VarState::Available { .. }) {
                Some(var)
            } else {
                None
            }
        })
        .join(", ");
    emql_error(
        38,
        usage.span(),
        format!("Invalid use of variable `{usage}`"),
    )
    .help(format!("Currently available variables are {vars}"))
}

pub fn query_let_variable_already_assigned(
    assign: &Ident,
    created: Span,
    used: Option<Span>,
) -> Diagnostic {
    let diag = emql_error(
        39,
        assign.span(),
        format!("Cannot assign to already created variable {assign}"),
    )
    .span_note(created, "Created here".to_string());
    if let Some(used) = used {
        diag.span_note(used, "Used here".to_string())
    } else {
        diag
    }
}

pub fn query_let_variable_shadows_table(assign: &Ident, table: &Ident) -> Diagnostic {
    emql_error(
        55,
        assign.span(),
        format!("variables created by let cannot shadow tables, but `{assign}` does"),
    )
    .span_note(table.span(), "Table defined here".to_string())
}

pub fn query_deref_cannot_deref_bag_type(
    lp: &Plan,
    reference: &Ident,
    t: &Key<RecordType>,
) -> Diagnostic {
    emql_error(
        40,
        reference.span(),
        format!(
            "Cannot dereference a bag of records `{}`",
            With {
                plan: lp,
                extended: t
            }
        ),
    )
}
pub fn query_cannot_return_stream(last: Span, ret: Span) -> Diagnostic {
    emql_error(41, ret, "Cannot return a stream from a query".to_string())
        .span_note(
            last,
            "The previous operator provides the values".to_string(),
        )
        .help("Use a `collect` operator to convert the stream into a bag of records".to_string())
}

pub fn query_table_access_nonexisted_columns(table_name: &Ident, col: &Ident) -> Diagnostic {
    emql_error(
        42,
        col.span(),
        format!("Cannot access {col} as it does not exist in {table_name}"),
    )
    .span_note(table_name.span(), format!("{table_name} defined here"))
}

pub fn query_invalid_record_type(
    lp: &Plan,
    op: &Ident,
    prev: Span,
    expected: &Key<RecordType>,
    found: &Key<RecordType>,
) -> Diagnostic {
    emql_error(
        43,
        op.span(),
        format!(
            "Data type does not match, expected {} but found {}",
            With {
                plan: lp,
                extended: expected
            },
            With {
                plan: lp,
                extended: found
            }
        ),
    )
}

pub fn query_no_cust_type_found(t: &Ident) -> Diagnostic {
    emql_error(44, t.span(), format!("Cannot find type {t}"))
}

pub fn table_query_no_such_field(table: &Ident, t: &Ident) -> Diagnostic {
    emql_error(45, t.span(), format!("no such field `{t}` in `{table}`"))
        .span_note(table.span(), format!("`{table}` defined here"))
}

pub fn query_cannot_append_to_record(new: &Ident, existing: &Ident) -> Diagnostic {
    emql_error(
        46,
        new.span(),
        format!("Cannot append new field `{new}` as it is already defined"),
    )
    .span_note(existing.span(), format!("{existing} defined here"))
}

pub fn sort_field_used_twice(field: &Ident, dup_field: &Ident) -> Diagnostic {
    emql_error(47, field.span(), format!("Field `{field}` is used twice in th sort order, sorts can only sort of each field once"))
    .span_note(dup_field.span(), format!("`{dup_field}` first used here"))
}

pub fn union_requires_at_least_one_input(call: &Ident) -> Diagnostic {
    emql_error(
        48,
        call.span(),
        format!("`{call}` requires at least one input"),
    )
}

pub fn operator_requires_streams(call: &Ident, var: &Ident) -> Diagnostic {
    emql_error(
        49,
        var.span(),
        format!("`{call}` inputs must be streams, but `{var}` is not a stream"),
    )
}

pub fn operator_requires_streams2(call: &Ident) -> Diagnostic {
    emql_error(50, call.span(), format!("`{call}` input must be a stream"))
}

pub fn no_return_in_context(call: &Ident) -> Diagnostic {
    emql_error(
        51,
        call.span(),
        format!("No return from `{call}` is present, needed for the output of `{call}`"),
    )
}

pub fn union_not_same_type(
    lp: &plan::Plan,
    call: &Ident,
    var: &Ident,
    data_type: &plan::Key<plan::RecordType>,
    other_var: &Ident,
    other_data_type: &plan::Key<plan::RecordType>,
) -> Diagnostic {
    emql_error(52, other_var.span(), format!("`{other_var}` has type `{}` but union requires all inputs to be of the same type `{}` (from `{var}`)", plan::With { plan: lp, extended: other_data_type }, plan::With { plan: lp, extended: data_type }))
}

pub fn query_deref_cannot_deref_table_get(
    lp: &plan::Plan,
    reference: &Ident,
    table: plan::Key<plan::Table>,
    field: &plan::RecordField,
) -> Diagnostic {
    let name = &lp.get_table(table).name;
    emql_error(53, reference.span(), format!("Cannot dereference a field directly taken from a table (this is `{name}.{field}`). Try mapping this into a table reference"))
    .note(format!("But what if `{name}.{field}` *is* a table reference? When dereferencing values from a table, the returned value is not the same type as the column, it can be optimised (for example it could be optimised into returning an Rc, a Cow, or a reference to the data). Rather than automatically copying out the value in these cases, it it left to the user to decide how they want to extract this."))
}

pub fn query_combine_extra_field(
    lp: &plan::Plan,
    call: &Ident,
    field: &Ident,
    data_type: &plan::Key<plan::RecordType>,
) -> Diagnostic {
    emql_error(54, field.span(), format!("Field `{field}` is not present in the type for `{call}` (same fields in input as output)"))
    .span_note(field.span(), format!("The type is: `{}`", plan::With { plan: lp, extended: data_type }))
}

pub fn query_combine_missing_field(
    lp: &plan::Plan,
    call: &Ident,
    field: &Ident,
    data_type: &plan::Key<plan::RecordType>,
) -> Diagnostic {
    emql_error(54, field.span(), format!("Field `{field}` is required but not present in the type for `{call}` (same fields in input as output)"))
    .span_note(field.span(), format!("The type is: `{}`", plan::With { plan: lp, extended: data_type }))
}

pub fn query_select_no_field(table: &plan::Table, field: &Ident) -> Diagnostic {
    let cols = table.columns.iter().filter_map(|(rf, _)| if let plan::RecordField::User(id) = rf { Some(id) } else {None}).join(", ");
    
    emql_error(55, field.span(), format!("Field `{field}` is not present in the table `{}`", table.name))
        .span_note(table.name.span(), format!("`{}` defined here with fields: {}", table.name, cols))
}

pub fn query_select_duplicate_field(field_original: &Ident, field: &Ident) -> Diagnostic {
    emql_error(56, field.span(), format!("Duplicate selection on {field}"))
        .span_note(field_original.span(), format!("`{field}` selected here"))
}