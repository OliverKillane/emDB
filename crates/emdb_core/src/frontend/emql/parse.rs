//! # Parsing
//! The main parser for the emql language
//! - Operator parsers are defined in [`super::operators`]
//! - Common parsing utilities are defined here
//! - The main [`emql_parser`] to invoke for emql through [`parse`].
//!
//! ## Design
//! Language design is discussed in [emql](super).
//!
//! These parsers rely on 1 token lookahead, this results in the language having
//! - heavy use of initialiser keywords
//! - uncomplex/simple grammar, with no ambiguity
//!
//! ## Potential Improvements
//! ### Parallel Parsing
//! Relying on changes to [combi] and the `!Send + !Sync` properties of [`TokenStream`].

use super::{ast, operators::parse_operator};
use combi::{
    core::{choice, mapsuc, nothing, recover, recursive, seq, seqdiff, setrepr, RecursiveHandle},
    derived::many0,
    logical::{not, or},
    macros::{choices, seqs},
    tokens::{
        basic::{
            collectuntil, getident, gettoken, isempty, matchident, matchpunct, peekident,
            peekpunct, recovgroup, syn, terminal,
        },
        derived::{listseptrailing, syntopunct},
        error::{embelisherr, error},
        recovery::until,
        TokenDiagnostic, TokenIter, TokenParser,
    },
    Combi,
};
use proc_macro2::{Delimiter, Ident, Span, TokenStream};
use proc_macro_error::{Diagnostic, Level};
use std::collections::LinkedList;
use syn::{Expr, Type};

pub(super) fn parse(ts: TokenStream) -> Result<ast::Ast, LinkedList<Diagnostic>> {
    let parser = emql_parser();
    let (_, res) =
        mapsuc(seqdiff(parser, terminal), |(o, ())| o).comp(TokenIter::from(ts, Span::call_site()));
    res.to_result().map_err(TokenDiagnostic::into_list)
}

enum EmqlItem {
    Query(ast::Query),
    Table(ast::Table),
    Backend(ast::BackendImpl),
}

fn emql_parser() -> impl TokenParser<ast::Ast> {
    mapsuc(
        many0(
            not(isempty()),
            recover(
                choices!(
                    peekident("query") => mapsuc(query_parser(), EmqlItem::Query),
                    peekident("table") => mapsuc(table_parser(), EmqlItem::Table),
                    peekident("impl") => mapsuc(backend_parser(), EmqlItem::Backend),
                    otherwise => error(gettoken, |t| {
                        Diagnostic::spanned(t.span(), Level::Error, String::from("expected impl, query or table"))
                    })
                ),
                until(or(
                    peekident("table"),
                    or(peekident("query"), peekident("impl")),
                )),
            ),
        ),
        |emql_items| {
            let mut tables = vec![];
            let mut queries = vec![];
            let mut backends = vec![];
            for obj in emql_items {
                match obj {
                    EmqlItem::Query(q) => queries.push(q),
                    EmqlItem::Table(t) => tables.push(t),
                    EmqlItem::Backend(b) => backends.push(b),
                }
            }
            ast::Ast {
                backends,
                tables,
                queries,
            }
        },
    )
}

fn backend_parser() -> impl TokenParser<ast::BackendImpl> {
    mapsuc(
        seq(
            recover(
                seqs!(
                    matchident("impl"),
                    getident(),
                    matchident("as"),
                    getident(),
                    choices!(
                        peekpunct(';') => mapsuc(nothing(), |()| None),
                        otherwise => mapsuc(recovgroup(Delimiter::Brace, collectuntil(isempty())), Option::Some)
                    )
                ),
                until(peekpunct(';')),
            ),
            matchpunct(';'),
        ),
        |((_, (impl_name, (_, (backend_name, options)))), _)| ast::BackendImpl {
            impl_name,
            backend_name,
            options,
        },
    )
}

pub type ContextRecurHandle =
    RecursiveHandle<TokenIter, TokenIter, Vec<ast::StreamExpr>, TokenDiagnostic, TokenDiagnostic>;

fn context_parser() -> impl TokenParser<Vec<ast::StreamExpr>> {
    recursive(|context_recur: ContextRecurHandle| {
        many0(
            not(isempty()),
            recover(
                recursive(|operator_recur| {
                    mapsuc(
                        seq(
                            parse_operator(context_recur.clone()),
                            choice(
                                peekpunct(';'),
                                mapsuc(matchpunct(';'), |_| None),
                                mapsuc(seq(connector_parse(), operator_recur), |(c, s)| {
                                    Some((c, Box::new(s)))
                                }),
                            ),
                        ),
                        |(op, con)| ast::StreamExpr { op, con },
                    )
                }),
                until(choice(
                    peekpunct(';'),
                    mapsuc(matchpunct(';'), |_| true),
                    mapsuc(nothing(), |()| false),
                )),
            ),
        )
    })
}

fn query_parser() -> impl TokenParser<ast::Query> {
    mapsuc(
        seqs!(
            matchident("query"),
            getident(),
            recovgroup(Delimiter::Parenthesis, query_param_list_parser()),
            recovgroup(Delimiter::Brace, context_parser())
        ),
        |(_, (name, (params, streams)))| ast::Query {
            name,
            context: ast::Context { params, streams },
        },
    )
}

fn table_parser() -> impl TokenParser<ast::Table> {
    mapsuc(
        seqs!(
            matchident("table"),
            getident(),
            recovgroup(Delimiter::Brace, member_list_parser()),
            choice(
                peekpunct('@'),
                mapsuc(
                    seq(
                        matchpunct('@'),
                        recovgroup(
                            Delimiter::Bracket,
                            listseptrailing(',', constraint_parser())
                        )
                    ),
                    |(_, cons)| cons
                ),
                mapsuc(nothing(), |()| vec![])
            )
        ),
        |(_, (name, (cols, cons)))| ast::Table { name, cols, cons },
    )
}

fn member_list_parser() -> impl TokenParser<Vec<(Ident, Type)>> {
    listseptrailing(
        ',',
        mapsuc(
            seqs!(getident(), matchpunct(':'), syntopunct(peekpunct(','))),
            |(m, (_, t))| (m, t),
        ),
    )
}

fn query_param_list_parser() -> impl TokenParser<Vec<(Ident, ast::AstType)>> {
    setrepr(
        listseptrailing(
            ',',
            mapsuc(
                seqs!(getident(), matchpunct(':'), type_parser_to_punct(',')),
                |(m, (_, t))| (m, t),
            ),
        ),
        "<name> : <Type>, ...",
    )
}

fn constraint_parser() -> impl TokenParser<ast::Constraint> {
    fn inner(
        name: &'static str,
        p: impl TokenParser<ast::ConstraintExpr>,
    ) -> impl TokenParser<ast::Constraint> {
        mapsuc(
            seqs!(
                matchident(name),
                recovgroup(Delimiter::Parenthesis, p),
                matchident("as"),
                getident()
            ),
            |(method, (p, (_, alias)))| ast::Constraint {
                alias,
                method_span: method.span(),
                expr: p,
            },
        )
    }

    choices!(
        peekident("unique") => inner("unique", mapsuc(getident(), |i| ast::ConstraintExpr::Unique{field:i})),
        peekident("pred") => inner("pred", mapsuc(syn(collectuntil(isempty())), ast::ConstraintExpr::Pred)),
        peekident("limit") => inner("limit", mapsuc(syn(collectuntil(isempty())), |e| ast::ConstraintExpr::Limit{size:e})),
        otherwise => error(getident(), |i| Diagnostic::spanned(i.span(), Level::Error, format!("expected a constraint (e.g. pred, unique) but got {i}")))
    )
}

fn connector_parse() -> impl TokenParser<ast::Connector> {
    embelisherr(
        choices!(
            peekpunct('~') => mapsuc(
                seq(
                    matchpunct('~'),
                    matchpunct('>')
                ),
                |(t1, _)| ast::Connector{stream: false, span: t1.span()}
            ),
            peekpunct('|') => mapsuc(
                seq(
                    matchpunct('|'),
                    matchpunct('>')
                ),
                |(t1, _)| ast::Connector{stream: true, span: t1.span()}
            ),
            otherwise => error(
                seq(
                    gettoken,
                    gettoken
                ),
                |(t1, t2)| Diagnostic::spanned(t1.span(), Level::Error, format!("expected either ~> or |> but got {t1}{t2}"))
            )
        ),
        "Connect operators a single row passed (`~>`), or a stream of rows (`|>`)",
    )
}

pub fn type_parser(end: impl TokenParser<bool>) -> impl TokenParser<ast::AstType> {
    setrepr(
        choices! {
            peekident("ref") => mapsuc(seq(matchident("ref"), getident()), |(_, i)| ast::AstType::TableRef(i)),
            peekident("type") => mapsuc(seq(matchident("type"), getident()), |(_, i)| ast::AstType::Custom(i)),
            otherwise => mapsuc(syntopunct(end), ast::AstType::RsType)
        },
        "<type>",
    )
}

pub fn type_parser_to_punct(end: char) -> impl TokenParser<ast::AstType> {
    type_parser(peekpunct(end))
}

// helper function for the functional style operators
pub fn functional_style<T>(
    name: &'static str,
    p: impl TokenParser<T>,
) -> impl TokenParser<(Ident, T)> {
    seq(matchident(name), recovgroup(Delimiter::Parenthesis, p))
}

pub fn fields_expr() -> impl TokenParser<Vec<(Ident, Expr)>> {
    listseptrailing(
        ',',
        mapsuc(
            seqs!(getident(), matchpunct('='), syntopunct(peekpunct(','))),
            |(id, (_, exp))| (id, exp),
        ),
    )
}

pub fn fields_assign() -> impl TokenParser<Vec<(Ident, (ast::AstType, Expr))>> {
    listseptrailing(
        ',',
        mapsuc(
            seqs!(
                setrepr(getident(), "<field name>"),
                matchpunct(':'),
                type_parser_to_punct('='),
                matchpunct('='),
                setrepr(syntopunct(peekpunct(',')), "<expression>")
            ),
            |(id, (_, (t, (_, e))))| (id, (t, e)),
        ),
    )
}
