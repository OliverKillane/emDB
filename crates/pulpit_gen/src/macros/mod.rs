pub mod raw_interface;

pub mod new_simple;
pub mod simple;
