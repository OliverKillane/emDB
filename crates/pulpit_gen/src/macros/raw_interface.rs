//! # Direct Control over tabl structure
//! A raw interface to allow users to choose each data type, for the

use proc_macro2::TokenStream;




/// Provides raw access to generate table structures
/// ```
/// raw_interface!{
///     primary: <column name> {
///         mut field: type,
///         field: type,
///     }
///     associated: [
///         <column name> {
///             mut field: type,
///             field: type,
///         }
///     ],
///     updates: [
///         
///     ],
///     deletions,
///     transactions,
/// }
/// ```
pub fn raw_interface(
    tks: TokenStream,
) {



} 

