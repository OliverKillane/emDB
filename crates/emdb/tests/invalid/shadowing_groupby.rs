use emdb::macros::emql;

emql!{
    table foos {
        key: i32,
        val: i32,
    }

    query shadow_time() {
        use foos
            |> groupby(key for let foos /* cannot shadow table here! */ in {
                row()
                    ~> return;
            });
    }
}

fn main() {}