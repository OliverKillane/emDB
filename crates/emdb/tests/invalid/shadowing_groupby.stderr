error: [EMQL-55] variables created by let cannot shadow tables, but `foos` does

         = note: Table defined here

  --> tests/invalid/shadowing_groupby.rs:11:36
   |
11 |             |> groupby(key for let foos /* cannot shadow table here! */ in {
   |                                    ^^^^
